### ISSUE TITLE
.ToList() and .ToDictionary() always considered as apparent type by csharp_style_var_when_type_is_apparent

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#76260]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/ToList-and-ToDictionary-always-con/10761277)._

---
We recently implemented a project-wide enforce for the following rules:
csharp_style_var_elsewhere = false:error
csharp_style_var_for_built_in_types = false:error
csharp_style_var_when_type_is_apparent = true:error

So I noticed the rule "csharp_style_var_when_type_is_apparent" now forces me to use 'var' for types, which aren't as apparent, specifically for all cases where we work with LINQ extension methods on Enumerable and we end them with ".ToList()" or ".ToDictionary()".
The code I'm including below shows a fairly basic class architecture, which allows me to show how types aren't apparent in the most basic LINQ cases, but are reported as apparent.

```
class Item(int id)
{
    public int ID { get; } = id;
}

class Data(int id)
{
    public int ID { get; } = id;
}

static class DataService
{
    public static Data LoadItem(Item item)
    {
        return new Data(item.ID);
    }
}

internal class Program
{
    private static void Main(string[] args)
    {
        List&lt;Item&gt; items = [new Item(1), new Item(2), new Item(3)];

        // no red underline, no need to use 'var' - as expected
        Data data = DataService.LoadItem(items[1]);

        // red underline for List&lt;Data&gt;, expecting me to change it to 'var'. this should not be the case, because DataService.LoadItem obfuscates the type
        List&lt;Data&gt; datasList = items
            .Where(i =&gt; i.ID &gt; 1 && i.ID &lt; 3)
            .Select(DataService.LoadItem)
            .ToList();

        // red underline for Dictionary&lt;int, Data&gt;, expecting me to change it to 'var'. this should not be the case, because DataService.LoadItem obfuscates the type
        // also d.ID for the key could be of any type - int, uint, custom struct, etc.
        Dictionary&lt;int, Data&gt; datasDictionary = items
            .Where(i =&gt; i.ID &gt; 1 && i.ID &lt; 3)
            .Select(DataService.LoadItem)
            .ToDictionary(d =&gt; d.ID);

        // no red underline - expected
        Data[] datasArray = items
            .Where(i =&gt; i.ID &gt; 1 && i.ID &lt; 3)
            .Select(DataService.LoadItem)
            .ToArray();

        // no red underline - expected
        IEnumerable&lt;Data&gt; datasIEnumerable = items
            .Where(i =&gt; i.ID &gt; 1 && i.ID &lt; 3)
            .Select(DataService.LoadItem);
    }
}
```

You can clearly see, that the types in all cases are not apparent, especially because we have a .Select() method in there, which changes the initial collection types, and also ending with a method "ToList()" or "ToDictionary()". And in general if we have a Select(), then it becomes very hard to know what would be the final type of the collection when calling a ToList.
I believe that both ToList and ToDictionary should not trigger the "apparent" flag and should not ask me to use 'var'.

---
#### Feedback Bot on 8/10/2024, 02:33 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>
